crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -279,66 +299,385 @@ where
+                }
+            }
+        }
+        Ok(None)
+    }
+
+    // Try to determine which address would cause query_ibc_denom to yield the
+    // supplied denom
+    fn reverse_query_ibc_denom(
+        denom: &PrefixedDenom,
+        ibc_denoms: &BTreeMap<String, Address>,
+    ) -> Option<Address> {
+        ibc_denoms
+            .get(&denom.to_string())
+            .cloned()
+            // If the reverse lookup failed, then guess the Address
+            // that might have yielded the IBC denom. However,
+            // guessing an IBC token address cannot possibly be
+            // correct due to the structure of query_ibc_denom
+            .or_else(|| {
+                Address::decode(denom.to_string()).ok().filter(|x| {
+                    !matches!(
+                        x,
+                        Address::Internal(InternalAddress::IbcToken(_))
+                    )
+                })
+            })
+    }
+
+    // Apply the given send packet to the changed balances structure
+    fn apply_send_packet(
+        &self,
+        mut acc: ChangedBalances,
+        msg: &MsgTransfer,
+        keys_changed: &BTreeSet<Key>,
+    ) -> Result<ChangedBalances> {
+        // If a key change with the same port, channel, and commitment as this
+        // message cannot be found, then ignore this message
+        if self
+            .search_ibc_transfer(&msg.message, keys_changed)?
+            .is_none()
+        {
+            return Ok(acc);
+        };
+
+        // Since IBC denominations are derived from Addresses
+        // when sending, we have to do a reverse look-up of the
+        // relevant token Address
+        let Some(token) = Self::reverse_query_ibc_denom(
+            &msg.message.packet_data.token.denom,
+            &acc.ibc_denoms,
+        ) else {
+            return Ok(acc);
+        };
+        let delta = ValueSum::from_pair(
+            token.clone(),
+            Amount::from_uint(Uint(*msg.message.packet_data.token.amount), 0)
+                .unwrap(),
+        );
+        // If there is a transfer to the IBC account, then deduplicate the
+        // balance increase since we already accounted for it above
+        if is_sender_chain_source(
+            msg.message.port_id_on_a.clone(),
+            msg.message.chan_id_on_a.clone(),
+            &msg.message.packet_data.token.denom,
+        ) {
+            let post_entry =
+                acc.post.entry(addr_taddr(IBC)).or_insert(ValueSum::zero());
+            *post_entry =
+                checked!(post_entry - &delta).map_err(native_vp::Error::new)?;
+        }
+        // Required for the packet's receiver to get funds
+        let post_entry = acc
+            .post
+            .entry(ibc_taddr(msg.message.packet_data.receiver.to_string()))
+            .or_insert(ValueSum::zero());
+        // Enable funds to be received by the receiver in the
+        // IBC packet
+        *post_entry =
```

Comment:
I guess the idea here is to mock a balance change to an address that we are not aware of on this chain (based on the ibc msg) so that then we can verify the matching between this ibc msg packet and masp transparent output with `validate_transparent_output`, correct?

Reply:
Also, are we validating the optional `ShieldingTransfer` that the receiving chain gets? I see that this message is built in `build_ibc_transfer` and we also link the hash of the unshielding masp section of the source chain. The method `execute` called by the ibc native vp returns this data but it seems like we never use it. This is probably a question for @yito88

Reply:
> I guess the idea here is to mock a balance change to an address that we are not aware of on this chain (based on the ibc msg) so that then we can verify the matching between this ibc msg packet and masp transparent output with `validate_transparent_output`, correct?

Yes, exactly.

Reply:
> The method execute called by the ibc native vp returns this data but it seems like we never use it.
Right, IBC actions never use `ShieldingTransfer`.

Reply:
@yito88 is this ok? Do we have a way to verify that the relayer is indeed forwarding the correct `ShieldingTransfer`?

Reply:
Ah, the IBC VP has to validate the `ShieldingTransfer` for the given IBC message.
I'll open another PR.